scripts/process/as_svg.R

```diff
@@ -0,0 +1,266 @@
+  deps <- readDepends(viz)
+  box <- deps[['clip_box']]
+  
+  elements <- clip_sf(deps[['data']], box) %>% 
+    mutate(.value = element_name) %>% 
+    select(.value, everything())
+  
+  append(list(elements = elements), plot_bounds(box))
+}
+
+#' takes output from one or many process.as_svg_{element} calls 
+#' and creates a `defs` svg container for it
+#' 
+#' @return writes an .rds file that contains the xml from `<g>` in a list
+#' format that can be converted to xml with xml2::as_xml_document()
+#' @details 
+#' if `depends` is empty, this still creates an element, it just 
+#' doesn't have xml children
+#' named `depends` are converted into `g` (group) names within `defs`
+process.as_svg_defs <- function(viz){
+  deps <- readDepends(viz)
+  defs <- xml_new_root('.x') 
+  group <- xml_add_child(defs, 'defs')
+  
+  for (g_name in names(deps)){
+    g <- xml_add_child(group, 'g', id = g_name)
+    svg_data <- deps[[g_name]]
+    for (j in seq_len(nrow(svg_data))){
+      do.call(xml_add_child, append(list(.x = g), svg_data[j, ]))
+    }
+  }
+  saveRDS(xml2::as_list(defs), file = viz[['location']])
+}
+
+#' create an xml/svg "group" or `<g>` element for each named dependency
+#' 
+#' @return writes an .rds file that contains the xml from `<g>` in a list
+#' format that can be converted to xml with xml2::as_xml_document()
+#' @details doesn't not name the id in the returned group 
+process.as_svg_g <- function(viz){
+  deps <- readDepends(viz)
+  g_main <- xml_new_root('.x') 
+  group <- do.call(xml_add_child, append(list(.x = g_main, .value = 'g'), viz[['attributes']]))
+  
+  for (g_name in names(deps)){
+    g <- xml_add_child(group, 'g', id = g_name)
+    svg_data <- deps[[g_name]]
+    for (j in 1:nrow(svg_data)){
+      do.call(xml_add_child, append(list(.x = g), svg_data[j, ]))
+    }
+  }
+
+  saveRDS(xml2::as_list(g_main), file = viz[['location']])
+}
+
+#' set up the basic plot par for a map
+set_svg_plot <- function(xlim, ylim){
+  par(mai=c(0,0,0,0), omi=c(0,0,0,0), xaxs = 'i', yaxs = 'i')
+  plot(NA, 0, xlim = xlim, ylim = ylim, axes = FALSE, frame.plot = FALSE, ann = FALSE)
+}
+
+
+#' get the xlim and ylim of a spatial object, if present. 
+#' Otherwise, return NULL
+plot_bounds <- function(sfc_object){
+  if (is.null(sfc_object)){
+    NULL
+  } else {
+    ppi <- 72
+    bounds <- sf::st_bbox(sfc_object)
+    svg_viewbox <- attr(sfc_object, 'svg_viewbox')
+    margins <- strsplit(svg_viewbox, '[ ]')[[1]] %>% as.numeric()
+    if (margins[1] != 0 | margins[2] != 0){
+      stop(call. = FALSE, 'the svg viewBox attribute must have xmin and ymin of zero')
+    }
+    list(xlim = bounds[c('xmin','xmax')], 
+         ylim = bounds[c('ymin','ymax')],
+         width = margins[3]/ppi,
+         height = margins[4]/ppi)
+  }
+}
+
+#' perform an `st_intersection` clip operation on sf object
+
+clip_sf <- function(sf_object, sf_clip = NULL){
+  if (!is.null(sf_clip)){
+    st_intersection(st_buffer(sf_object, 0), sf_clip)
```

Comment:
add `sf::`

Reply:
I might want to bring this up as a question for the style-guide team. xml2 and sf both do this prefixing thing, where it is pretty clear where the functions come from (`st_` and `xml_`). do we need the redundancy of `xml2::xml_{x}`?

Reply:
I don't see it as redundancy. I wouldn't necessarily assume that _all_ functions prefixed with `st_`or `xml_` are from `sf` and `xml2` or that everyone would know to make that assumption.

Reply:
I'm updating this, but I disagree with the need for me to add `xml2::xml_` to a bunch of things when above you are OK with me leaving the `dplyr` verbs alone, which aren't prefixed. 

But no, I don't want to add `dplyr::` to every line in the munges either...